File: game_engine/message_codes.py
"""Stable message codes for game engine results (UI text lives on the client)."""

# Turn / move flow
TURN_NOW = "turn.now"
MOVE_PASSED = "move.passed"
MOVE_ILLEGAL = "move.illegal"
MOVE_NO_PIECE = "move.no_piece"
MOVE_WRONG_COLOR = "move.wrong_color"
MOVE_TARGET_OCCUPIED = "move.target_occupied"
MOVE_OWN_PIECE_BLOCKS = "move.own_piece_blocks"
MOVE_UNKNOWN_PIECE = "move.unknown_piece"
MOVE_INVALID_EVENT = "move.invalid_event"
MOVE_NO_CAPTURE_TARGET = "move.no_capture_target"

# Capture chain
CAPTURE_CONTINUE = "capture.continue"
CAPTURE_CONTINUE_SAME = "capture.continue_same"
CAPTURE_MUST = "capture.must"
CAPTURE_MUST_CONTINUE = "capture.must_continue"
CAPTURE_MANDATORY_OTHER = "capture.mandatory_other"
CAPTURE_ONLY_BIY = "capture.only_biy"

# Promotion / endgame
PIECE_PROMOTED = "piece.promoted"
GAME_DRAW = "game.draw"

# Validation code -> message code
VALIDATION_TO_MESSAGE = {
    "NO_PIECE": MOVE_NO_PIECE,
    "WRONG_COLOR": MOVE_WRONG_COLOR,
    "TARGET_OCCUPIED": MOVE_TARGET_OCCUPIED,
    "MANDATORY_CAPTURE_OTHER_PIECE": CAPTURE_MANDATORY_OTHER,
    "ONLY_BIY_CAN_CAPTURE": CAPTURE_ONLY_BIY,
    "MANDATORY_CAPTURE_THIS_PIECE": CAPTURE_MUST,
    "OWN_PIECE_BLOCKS_BATYR": MOVE_OWN_PIECE_BLOCKS,
    "ILLEGAL_MOVE": MOVE_ILLEGAL,
    "INTERNAL_NO_PIECE": MOVE_NO_PIECE,
    "OK_CAPTURE": "",
    "OK_MOVE": "",
}

# Draw reasons (stored in game.reason)
DRAW_TWO_BIYS = "draw_two_biys"
DRAW_REPETITION = "draw_repetition"

File: crates/shatra-engine/src/rules/message_codes.rs
pub const TURN_NOW: &str = "turn.now";
pub const MOVE_PASSED: &str = "move.passed";
pub const MOVE_ILLEGAL: &str = "move.illegal";
pub const MOVE_NO_PIECE: &str = "move.no_piece";
pub const MOVE_WRONG_COLOR: &str = "move.wrong_color";
pub const MOVE_TARGET_OCCUPIED: &str = "move.target_occupied";
pub const MOVE_OWN_PIECE_BLOCKS: &str = "move.own_piece_blocks";
pub const MOVE_UNKNOWN_PIECE: &str = "move.unknown_piece";
pub const CAPTURE_CONTINUE: &str = "capture.continue";
pub const CAPTURE_CONTINUE_SAME: &str = "capture.continue_same";
pub const CAPTURE_MUST: &str = "capture.must";
pub const CAPTURE_MUST_CONTINUE: &str = "capture.must_continue";
pub const CAPTURE_MANDATORY_OTHER: &str = "capture.mandatory_other";
pub const CAPTURE_ONLY_BIY: &str = "capture.only_biy";
pub const PIECE_PROMOTED: &str = "piece.promoted";

pub fn validation_to_message(code: &str) -> &'static str {
    match code {
        "NO_PIECE" | "INTERNAL_NO_PIECE" => MOVE_NO_PIECE,
        "WRONG_COLOR" => MOVE_WRONG_COLOR,
        "TARGET_OCCUPIED" => MOVE_TARGET_OCCUPIED,
        "MANDATORY_CAPTURE_OTHER_PIECE" => CAPTURE_MANDATORY_OTHER,
        "ONLY_BIY_CAN_CAPTURE" => CAPTURE_ONLY_BIY,
        "MANDATORY_CAPTURE_THIS_PIECE" => CAPTURE_MUST,
        "OWN_PIECE_BLOCKS_BATYR" => MOVE_OWN_PIECE_BLOCKS,
        "ILLEGAL_MOVE" => MOVE_ILLEGAL,
        _ => MOVE_ILLEGAL,
    }
}
